cmd/elementGen/main.go

```diff
@@ -0,0 +1,745 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"go/format"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"text/template"
+)
+
+const (
+	primaryTemplate = `
+// Copyright (c) 2018 Paul Jolly <paul@myitcv.org.uk>, all rights reserved.
+// Use of this document is governed by a license found in the LICENSE document.
+
+package react
+
+// {{ .Elem }} is the React element definition corresponding to the HTML <{{ .Name }}> element.
+type {{ .Elem }} struct {
+	Element
+}
+
+// _{{ .Props }} defines the properties for the <{{ .Name }}> element.
+type _{{ .Props }} struct {
+	*BasicHTMLElement
+
+	{{ range .Attrs }}{{ .Name }} {{ .Type }} ` + "`js:\"{{ .JS }}\"`" + `
+	{{ end }}
+}
+
+// A creates a new instance of a <{{ .Name }}> element with the provided props and children.
+func {{ .Upper }}(props *{{ .Props }}, children ...Element) *{{ .Elem }} {
+	rProps := &_{{ .Props }}{
+		BasicHTMLElement: newBasicHTMLElement(),
+	}
+
+	if props != nil {
+		props.assign(rProps)
+	}
+
+	return &{{ .Elem }}{
+		Element: createElement("{{ .Name }}", rProps, children...),
+	}
+}
+`
+	testTemplate = `
+// +` + "build js" + `
+
+package react_test
+
+import (
+	"testing"
+
+	"honnef.co/go/js/dom"
+
+	"myitcv.io/react"
+	"myitcv.io/react/testutils"
+)
+
+func Test{{ .Elem }}(t *testing.T) {
+	class := "test"
+
+	x := testutils.Wrapper(react.{{ .Upper }}(&react.{{ .Props }}{ClassName: class}))
+	cont := testutils.RenderIntoDocument(x)
+
+	el := testutils.FindRenderedDOMComponentWithClass(cont, class)
+
+	if _, ok := el.(*dom.{{ .DOM }}); !ok {
+		t.Fatal("Failed to find <{{ .Name }}> element")
+	}
+}
+`
+)
+
+type (
+	Desc struct {
+		Override   string
+		Attributes []Attr
+		Basic      bool // True if honnef.co/go/js/dom declares a specific Element type.
+	}
+
+	Attr struct {
+		Name     string
+		Override string
+		Type     string
+	}
+
+	templElem struct {
+		Elem, Name, Props, Upper, DOM string
+		Attrs                         []templAttr
+	}
+
+	templAttr struct {
+		Name, JS, Type string
+	}
+)
+
+var (
+	outputDirectory = flag.String("o", ".", "output directory to write the generated Go files")
+
+	// elements contains all of the Go wrappers to generate for the underlying HTML elements.
+	// Commented items have already been hand-written.
+	elements = map[string]Desc{
```

Comment:
I'd put this variable declaration into a separate file in this package, `elements.go` or similar; keeps the "main" file short